File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        int[] newHeights = new int[n + 2];
        System.arraycopy(heights, 0, newHeights, 1, n); 
        n = newHeights.length;

        Stack<Integer> stack = new Stack<>();
        int maxArea = 0;

        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && newHeights[i] < newHeights[stack.peek()]) {
                int height = newHeights[stack.pop()];
                int width = i - stack.peek() - 1;
                maxArea = Math.max(maxArea, height * width);
            }
            stack.push(i);
        }

        return maxArea;
    }
}

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        stack = [] 
        max_area = 0
        heights.append(0)  
        
        for i in range(len(heights)):
            while stack and heights[i] < heights[stack[-1]]:
                h = heights[stack.pop()]
 
                left = stack[-1] if stack else -1
                width = i - left - 1
                
                max_area = max(max_area, h * width)
            
            stack.append(i)
        
        return max_area